File: Realize/3Week/PracticeHashtable.py
class MyHashTable:
    def __init__(self, size):
        self.size = size
        self.table = [None] * size
        self.elemCnt = 0

    # 해시 함수
    def get_hash(self, key):
        return key % self.size

    # 값 설정
    def set_value(self, key, data):
        idx = self.get_hash(key)
        self.table[idx] = data
        self.elemCnt += 1

    # 값 가져오기
    def get_value(self, key):
        idx = self.get_hash(key)
        return self.table[idx]

    # 값 삭제
    def remove_value(self, key):
        idx = self.get_hash(key)
        self.table[idx] = None
        self.elemCnt -= 1

    # 해시 테이블 출력
    def print_hash_table(self):
        print("== Hash Table ==")
        for i in range(self.size):
            print(f"{i}: {self.table[i] if self.table[i] is not None else 'null'}")


# 테스트 코드
if __name__ == "__main__":
    ht = MyHashTable(7)
    ht.set_value(1, 1)
    ht.set_value(2, 2)
    ht.set_value(3, 3)
    ht.set_value(4, 4)
    ht.set_value(5, 5)
    ht.print_hash_table()

    ht.set_value(8, 6)  # 충돌이 발생할 수 있음
    ht.print_hash_table()

File: Realize/3Week/PracticeHashtable.js
class MyHashTable {
    constructor(size) {
        this.size = size;
        this.table = new Array(size).fill(null);
        this.elemCnt = 0;
    }

    // 해시 함수
    getHash(key) {
        return key % this.size;
    }

    // 값 설정
    setValue(key, data) {
        const idx = this.getHash(key);
        this.table[idx] = data;
        this.elemCnt++;
    }

    // 값 가져오기
    getValue(key) {
        const idx = this.getHash(key);
        return this.table[idx];
    }

    // 값 삭제
    removeValue(key) {
        const idx = this.getHash(key);
        this.table[idx] = null;
        this.elemCnt--;
    }

    // 해시 테이블 출력
    printHashTable() {
        console.log("== Hash Table ==");
        for (let i = 0; i < this.size; i++) {
            console.log(`${i}: ${this.table[i] !== null ? this.table[i] : 'null'}`);
        }
    }
}

// 테스트 코드
const ht = new MyHashTable(7);
ht.setValue(1, 1);
ht.setValue(2, 2);
ht.setValue(3, 3);
ht.setValue(4, 4);
ht.setValue(5, 5);
ht.printHashTable();

ht.setValue(8, 6); // 충돌이 발생할 수 있음
ht.printHashTable();
